tree/forest/v7/inc/ROOT/RPage.hxx

```diff
@@ -0,0 +1,126 @@
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RPage
+\ingroup Forest
+\brief A page is a fixed size slice of a column that is mapped into memory
+
+The page provides a fixed-size opaque memory buffer for uncompressed data. It does not know how to interpret
+the contents but it does now about the size (and thus the number) of the elements inside as well as the element
+number range within the backing column. The memory buffer is not managed by the page but normally by the page pool.
+*/
+// clang-format on
+class RPage {
+public:
+   /**
+    * Stores information about the cluster in which this page resides.
+    */
+   class RClusterInfo {
+   private:
+      /// The cluster number
+      ForestSize_t fId;
+      /// The first element index of the column in this cluster
+      ForestSize_t fSelfOffset;
+      /// For offset columns, also store the cluster offset of the column being referenced
+      ForestSize_t fPointeeOffset;
+   public:
+      RClusterInfo() : fId(0), fSelfOffset(0), fPointeeOffset(0) {}
+      RClusterInfo(ForestSize_t id, ForestSize_t selfOffset, ForestSize_t pointeeOffset)
+         : fId(id), fSelfOffset(selfOffset), fPointeeOffset(pointeeOffset) {}
+      ForestSize_t GetId() const { return fId; }
+      ForestSize_t GetSelfOffset() const { return fSelfOffset; }
+      ForestSize_t GetPointeeOffset() const { return fPointeeOffset; }
+   };
+
+private:
+   ColumnId_t fColumnId;
+   void* fBuffer;
+   std::size_t fCapacity;
+   std::size_t fSize;
+   std::size_t fElementSize;
+   ForestSize_t fRangeFirst;
+   ForestSize_t fNElements;
+   RClusterInfo fClusterInfo;
+
+public:
+   RPage()
+     : fColumnId(kInvalidColumnId), fBuffer(nullptr), fCapacity(0), fSize(0), fElementSize(0),
+       fRangeFirst(0), fNElements(0)
+   {}
+   RPage(ColumnId_t columnId, void* buffer, std::size_t capacity, std::size_t elementSize)
+      : fColumnId(columnId), fBuffer(buffer), fCapacity(capacity), fSize(0), fElementSize(elementSize),
+        fRangeFirst(0), fNElements(0)
+      {}
+   ~RPage() = default;
+
+   std::int64_t GetColumnId() { return fColumnId; }
+   /// The total space available in the page
+   std::size_t GetCapacity() const { return fCapacity; }
+   /// The space taken by column elements in the buffer
+   std::size_t GetSize() const { return fSize; }
+   ForestSize_t GetNElements() const { return fSize / fElementSize; }
+   ForestSize_t GetRangeFirst() const { return fRangeFirst; }
+   ForestSize_t GetRangeLast() const { return fRangeFirst + fNElements - 1; }
+   const RClusterInfo& GetClusterInfo() const { return fClusterInfo; }
+   bool Contains(ForestSize_t index) const {
+      return (index >= fRangeFirst) && (index < fRangeFirst + fNElements);
+   }
+   void* GetBuffer() const { return fBuffer; }
+   /// Return a pointer after the last element that has space for nElements new elements. If there is not enough capacity,
+   /// return nullptr
+   void* Reserve(std::size_t nElements) {
```

Comment:
This is an unusual sematic for 'reverse' (see https://en.cppreference.com/w/cpp/container/vector/reserve).
Don't we mean 'Resize' here?  [blocker]

Reply:
Philippe means reserve, not reverse.

But the functionality sounds like "TryGrow"?